bench/macro/Bench/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -0,0 +1,494 @@
+--
+-- * where the same key is being looked up in many filters,
+-- * with a hit in only one filter expected, and
+-- * where the total size of the filters is too large to fully fit in cache
+--   (though the smaller ones may fit in the caches).
+--
+prepLookupsEnv ::
+     [RunSizeInfo]
+  -> StdGen -- ^ Key RNG
+  -> IO ( V.Vector (Run (FS.Handle FS.HandleIO))
+        , V.Vector (Bloom SerialisedKey)
+        , V.Vector IndexCompact
+        , V.Vector (FS.Handle FS.HandleIO)
+        , FilePath -- temporary directory that should be removed eventually
+        , FS.HasFS IO FS.HandleIO
+        , FS.HasBlockIO IO FS.HandleIO
+        )
+prepLookupsEnv runSizes keyRng0 = do
+    sysTmpDir <- getCanonicalTemporaryDirectory
+    benchTmpDir <- createTempDirectory sysTmpDir "lookups-macro-bench"
+    let hfs = FS.ioHasFS (FS.MountPoint benchTmpDir)
+    hbio <- FS.ioHasBlockIO hfs FS.defaultIOCtxParams
+
+    -- create the vector of initial keys
+    (mvec :: VUM.MVector RealWorld UTxOKey) <- VUM.unsafeNew (totalNumEntries runSizes)
+    !keyRng1 <- vectorOfUniforms mvec keyRng0
+    -- we reuse keyRng0 to generate batches of lookups, so by shuffling the
+    -- vector we ensure that these batches of lookups will do random disk
+    -- access.
+    !_ <- shuffle mvec keyRng1
+
+    -- create the runs
+    rbs <- sequence
+            [ RunBuilder.new hfs (RunFsPaths i) (NumEntries numEntries) (numEntries * 42)
+            | ((numEntries, _), i) <- zip runSizes [0..] ]
+
+    -- fill the runs
+    putStr "addKeyOp"
+    let zero = serialiseValue zeroUTxOValue
+    foldM_
+      (\ !i (!rb, !n) -> do
+        let !mvecLocal = VUM.unsafeSlice i n mvec
+        Merge.sort mvecLocal
+        forM_ [0..n-1] $ \j -> do
+          -- progress
+          when (j .&. 0xFFFF == 0) (putStr ".")
+          !k <- mvecLocal `VUM.unsafeRead` j
+          void $ RunBuilder.addKeyOp hfs rb (serialiseKey k) (Insert zero)
+        pure (i+n)
+      )
+      0
+      (zip rbs (fmap fst runSizes))
+    putStr "DONE"
+
+    -- return runs
+    runs <- V.fromList <$> mapM (Run.fromMutable hfs (Run.RefCount 1)) rbs
+    let blooms = V.map Run.runFilter runs
+        indexes = V.map Run.runIndex runs
+        handles = V.map Run.runKOpsFile runs
+    pure $!! (runs, blooms, indexes, handles, benchTmpDir, hfs, hbio)
+
+genLookupBatch :: StdGen -> Int -> (V.Vector SerialisedKey, StdGen)
+genLookupBatch !rng0 !n0
+  | n0 <= 0   = error "mkBatch: must be positive"
+  | otherwise = runST $ do
+      mres <- VM.unsafeNew n0
+      go rng0 0 mres
+  where
+    go ::
+         StdGen -> Int -> VM.MVector s SerialisedKey
+      -> ST s (V.Vector SerialisedKey, StdGen)
+    go !rng !i !mres
+      | n0 == i   = do
+          !res <- V.unsafeFreeze mres
+          pure (res, rng)
+      | otherwise = do
+          let (!k, !rng') = uniform @StdGen @UTxOKey rng
+              !sk = serialiseKey k
+          VM.write mres i $! sk
+          go rng' (i+1) mres
```

Comment:
Move to utilities, possibly also rework using `V.generateM`

Reply:
I recall now that I wrote it this way because I needed `StdGen` as part of the result of this function. `V.unfoldrExactN` doesn't return that